crates/matrix-sdk/src/authentication/service.rs

```diff
@@ -0,0 +1,566 @@
+            oidc_static_registrations,
+            #[cfg(feature = "experimental-sliding-sync")]
+            custom_sliding_sync_proxy: RwLock::new(custom_sliding_sync_proxy),
+        }
+    }
+
+    /// Returns the homeserver details for the currently configured homeserver,
+    /// or `None` if a successful call to `configure_homeserver` is yet to be
+    /// made.
+    pub fn homeserver_details(&self) -> Option<HomeserverLoginDetails> {
+        self.homeserver_details.read().unwrap().clone()
+    }
+
+    /// Updates the service to authenticate with the homeserver for the
+    /// specified address.
+    pub async fn configure_homeserver(
+        &self,
+        server_name_or_homeserver_url: String,
+    ) -> Result<(), AuthenticationError> {
+        let mut builder = self.new_client_builder();
+
+        // Attempt discovery as a server name first.
+        let result = sanitize_server_name(&server_name_or_homeserver_url);
+
+        match result {
+            Ok(server_name) => {
+                if server_name_or_homeserver_url.starts_with("http://") {
+                    builder = builder.insecure_server_name_no_tls(&server_name);
+                } else {
+                    builder = builder.server_name(&server_name);
+                }
+            }
+
+            Err(e) => {
+                // When the input isn't a valid server name check it is a URL.
+                // If this is the case, build the client with a homeserver URL.
+                if Url::parse(&server_name_or_homeserver_url).is_ok() {
+                    builder = builder.homeserver_url(server_name_or_homeserver_url.clone());
+                } else {
+                    return Err(e.into());
+                }
+            }
+        }
+
+        let client = match builder.build().await {
+            Ok(client) => Ok(client),
+            Err(e) => {
+                if !server_name_or_homeserver_url.starts_with("http://")
+                    && !server_name_or_homeserver_url.starts_with("https://")
+                {
+                    Err(e)
+                } else {
+                    // When discovery fails, fallback to the homeserver URL if supplied.
+                    let mut builder = self.new_client_builder();
+                    builder = builder.homeserver_url(server_name_or_homeserver_url);
+                    builder.build().await
+                }
+            }
+        }
+        .map_err(|e| AuthenticationError::Generic { message: e.to_string() })?;
+
+        let details = self.details_from_client(&client).await?;
+
+        // Now we've verified that it's a valid homeserver, make sure
+        // there's a sliding sync proxy available one way or another.
+        #[cfg(feature = "experimental-sliding-sync")]
+        if self.custom_sliding_sync_proxy.read().unwrap().is_none()
+            && client.sliding_sync_proxy().is_none()
+        {
+            return Err(AuthenticationError::SlidingSyncNotAvailable);
+        }
```

Comment:
Indeed, we can't mandate usage of sliding sync for users of the authentication API, so this would probably need to go. This was specific to the FFI layer which *only* uses Sliding Sync, so there it would make sense to have such a check IMO.